### ISSUE TITLE
Propagate `theme` attribute to themable parts

### ISSUE THREAD

**Issue by manolo** [vaadin/vaadin-themable-mixin#12]
The idea is to change the API so as themes added/removed in the component are propagated to every children that extends the themable-mixin.

With this solution for example we could have different date-pickers in the same page where each one have different appearance for the vaadin-month-calendar, vaadin-button, etc.

**Comment by jouni** [comment]
I’m leaning towards not supporting this idea.

While it solves some problems that are easy to spot, like having the same `small` theme for a normal `vaadin-text-field` and the text field inside `vaadin-date-picker`, it has some potentially problematic side-effects at the same time. For example, `<vaadin-date-picker theme="small">` would end up making the buttons in the overlay toolbar small as well, which could be unexpected and unwanted.

Also, this approach doesn’t solve other related problems with the `theme` attribute, like how to use a certain element theme when it’s embedded inside another element. For example, I want to use the "tertiary" theme for the date picker overlay buttons, but I can’t since I can’t set the theme property using another theme module.

**Comment by jouni** [comment]
Continuing brainstorming here:

When an element (with `ThemableMixin`) is attached, it reads the `--theme` property from computed style, and appends those to the `theme` property/attribute.

The "small" style text field case could then be solved like this:

In date picker theme:
```css
:host([theme~="small"]) [part="text-field"] {
  --theme: "small";
}
```

And The overlay buttons case could be solved like this:

In date picker overlay theme:
```css
[part="toolbar"] [part$="button"] {
  --theme: "tertiary";
}
```

**Comment by jouni** [comment]
The use cases I outlined above:

1. Have the same "small" theme for a normal `vaadin-text-field` and the text field inside `vaadin-date-picker` or `vaadin-combo-box`, so the following elements would have the same appearence:

```html
<vaadin-text-field theme="small"></vaadin-text-field>
<vaadin-date-picker theme="small"></vaadin-date-picker>
<vaadin-combo-box theme="small"></vaadin-combo-box>
```

**Comment by platosha** [comment]
> ```
> :host([theme~="small"]) [part="text-field"] {
>   --theme: "small";
> }
> ```

I’m pretty sure ScopingShim from ShadyCSS wouldn’t recognize this. But something like `host.updateStyle({'--theme': host.getAttribute('theme')});` would apply `--theme` to the entire shadow root.

**Comment by manolo** [comment]
At the same time it would be nice if we are able to change the theme at runtime. 
probably `theme` attribute could have an observer, and when the user changes it `datepicker.theme = 'large'`  a method `updateStyles` which is called as well in all nested themable elements.

**Comment by jouni** [comment]
> I’m pretty sure ScopingShim from ShadyCSS wouldn’t recognize this.

Actually, it does. I quickly tested this, and I can read the correct value of the property with `ShadyCSS.getComputedStyleValue()` in `vaadin-date-picker`’s `ready` method, using that exact same selector in a style module. Tested with EdgeHTML 14.14393.

**Comment by jouni** [comment]
One approach to solving use case #1, which @Legioth suggested, was to have generated attributes/properties for applying themes for the named parts inside an element. 

For example, for the date picker which has a `[part="text-field"]`, the user could apply a theme for it like this:

```html
<vaadin-date-picker text-field-theme="small"></vaadin-date-picker>
```

Seems like this would be quite flexible and give a good amount of power to the developer. But it also requires the developer to know the types and possible theme names of the embedded elements. That would most likely make it more difficult to find out all the possibilities for one single element.

Where it fails, though, is if the element is nested more deeply in shadow roots. For example, let’s say a developer has created a new composite element which wraps two `vaadin-date-picker`’s, which they have named as `part="date-picker-start"` and `part="date-picker-end"`. How does the user of that element apply the "small" theme for those date pickers? `<date-range date-picker-start-text-field="small" date-picker-end-text-field="small">`? Becomes verbose very quickly, and requires again the developer to know a lot of the internals of the element.

Comparing to the option of using a `--theme` CSS property, which gives the power to the theme developer to define all the possible `theme` attribute values an element supports and add them to the documentation of that element.

**Comment by jouni** [comment]
With the `--theme` approach, to prevent it from inheriting by default, all element themes should specify `--theme: var(--element-name-theme, initial);` for the `:host` element. This will avoid similar cases of themes being applied unexpectedly and unwantedly as with the original proposed appoach where the `theme` property would be propagated to all elements.

This would make it more controllable how the themes are propagated through shadow roots. The earlier examples would still work, but it would also open up the possibility to define the date picker overlay button themes from the date picker theme (vs. from the date picker overlay theme):

In date picker theme:
```css
[part="overlay"] {
  --vaadin-button-theme: "tertiary";
}
```

This would also work similarly for the more nested use case of a `date-range` element. The author of that element could expose `theme="small"` for the `date-range` element which then applies the correct themes for the internal parts:

```html
<dom-module id="date-range-theme" theme-for="date-range">
  <template>
    <style>
      :host([theme~="small"]) [part^="date-picker"] {
        --vaadin-text-field-theme: "small";
      }
    </style>
  </template>
</dom-module>
```

**Comment by manolo** [comment]
There is a performance regression with this approach as described in #17.

The --theme approach needs to run the `getComputedStyle.getPropertyValue` that forces reflow.

Possible solutions would be to prevent the usage of themable-mixin elements in large doms, not use css custom properties but attributes like  `theme-for-part-input="foo"`

**Comment by jouni** [comment]
I don’t see how attributes can solve all the use cases outlined above. Could you provide examples?

**Comment by platosha** [comment]
Research from vaadin/vaadin-dialog#37

TL/DR I propose basically the same as the solution by @manolo but with a helper mixin:
- add a small ThemeProperty mixin: https://github.com/vaadin/vaadin-themable-mixin/blob/09aa2fb158624ffd71a74683e92a2f82e3f55707/vaadin-theme-property-mixin.html
- and use it in the wrapper components: https://github.com/vaadin/vaadin-dialog/pull/106/files

### Problem

Developers have no way to set the `theme` attribute for internal components. This concerns the following internal and wrapper components:

Internal component | Used inside
---|---
`<...-overlay>` | `<vaadin-context-menu>`, `<vaadin-combo-box>`, `<vaadin-date-picker>`, `<vaadin-dialog>`, `<vaadin-dropdown-menu>`, `<vaadin-time-picker>`
`<-text-field>` | `<vaadin-combo-box>`, `<vaadin-date-picker>`, `<vaadin-dropdown-menu>`, `<vaadin-grid-filter>`, `<vaadin-time-picker>`
`<-card>` | `<vaadin-notification>`
`<-file>` | `<vaadin-upload>` (no builtin theme variants yet)

Ideally, the solution should be applicable to all of the above, and should not have side effects for other cases of the `theme` attribute usage.

Notes:
1. Consider that as a solution to the problem we introduce some way of propagating the `theme` value. In general, the propagated value might conflict with the value set directly in the attribute before the propagation. Fortunately, it would not happen in case if the propagation is scoped to the internal components use case only.
2. `<vaadin-text-field>` is both an internal and a public component itself.
3. All internal components extend `Vaadin.ThemableMixin`, but some wrapper components don’t, for ex., `<vaadin-dialog>`.

### Solution Ways

#### Way 1: internal components discover the theme from parent scope

Advantage: potentially might be implemented with changes to `Vaadin.ThemableMixin` alone, without changes to the parent component.

Disadvantages:
- Not well scoped to the problem. In `<vaadin-text-field>` case, it would be hard to distinguish the internal use from the public one, and the solution might have undesired side effects for the public case. For example, might produce a value conflict, as noted in the Note 1 above.
- Risk of dynamic value issues. It would be hard to support a dynamic `theme` update propagation from the parent. If done dynamically, `<vaadin-overlay>` might loose its theme when it is moved to the body on open.
- If done statically, there is a timing issues risk. The wrapper component has to define the `theme` value to propagate before the internal discovers it.
- It is tempting to use CSS Custom Properties for propagation, since they nicely inherit from the parent by design. However, this was already done once and then removed because of performance issues.

#### Way 2: the wrapper component actively defines the `theme` attribute values on the internal ones

The advantages are same as the disadvantages of the way 1:
- Well scoped, no side effects beyond the wrapper components implementing this solution.
- Could easily dynamically reflect the changes of the wrapper theme. No risk of loosing the overlay scope, the wrapper stores references to the overlay even when it is open and disconnected.
- Does not have to be static, the dynamic solution is easy.

Disadvantage: requires some implementation in the wrapper component.

### Proposal

Let us discard the way 1 and use the way 2.

Using explicit template bindings, e. g., `theme$="[[theme]]"` on the internal components, seems to be the easiest implementation of the way 2.

A static solution would be strange: the `theme` attribute on the simple components (think of `<vaadin-button>`) is naturally dynamic. So I assume the theme attribute propagation is expected to be dynamic too.

To facilitate the bindings and make the solution clean and dynamic, we have to define the theme property and set it from the attribute change callback with the `theme` attribute value. Because the same has to be done in many wrapper components, it makes sense to share that with a mixin.

`Vaadin.ThemableMixin` is not the best for sharing the theme property, because, as Note 3 says, not every wrapper component extend it. Therefore, a separate mixin. We can use the same `vaadin-themable-mixin` repository though, because it is anyways downloaded and used in the internal components.

See the prototype ThemePropertyMixin in this PR: https://github.com/vaadin/vaadin-themable-mixin/pull/32/files

The usage example in `<vaadin-dialog>` in this PR: https://github.com/vaadin/vaadin-dialog/pull/106/files

**Comment by web-padawan** [comment]
@platosha is this completed, or something is still missing? 
It looks like `time-picker` is in your list but has not been updated.

**Comment by web-padawan** [comment]
Time-picker added, so closing this now.
